File: sdks/python/src/bigrag/resources/admin/realtime.py
from __future__ import annotations

from collections.abc import AsyncGenerator
from typing import TYPE_CHECKING, Any

from bigrag._realtime import RealtimeConnection
from bigrag.types.admin import AdminRealtimeEvent

if TYPE_CHECKING:
    from bigrag._core import BigRAGCore


class AdminRealtimeResource:
    def __init__(self, client: BigRAGCore) -> None:
        self._client = client

    def documents(
        self,
        collection: str,
        *,
        status: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
        order: str | None = None,
        q: str | None = None,
        sort: str | None = None,
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream(
            "admin.collections.documents",
            {
                "collection": collection,
                "status": status,
                "limit": limit,
                "offset": offset,
                "order": order,
                "q": q,
                "sort": sort,
            },
        )

    def document_batch_status(
        self, collection: str, document_ids: list[str]
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream(
            "admin.collections.documents.batch_status",
            {"collection": collection, "document_ids": document_ids},
        )

    def document(
        self, collection: str, document_id: str
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream(
            "admin.collections.documents.detail",
            {"collection": collection, "document_id": document_id},
        )

    def upload_session(
        self, collection: str, session_id: str
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream(
            "admin.collections.upload_session",
            {"collection": collection, "session_id": session_id},
        )

    def collection_stats(self, collection: str) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream("admin.collections.stats", {"collection": collection})

    def connector_sources(
        self, provider: str, *, collection: str | None = None
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream(
            "admin.connectors.sources",
            {"provider": provider, "collection": collection},
        )

    def connector_sync_jobs(
        self,
        provider: str,
        *,
        collection: str | None = None,
        source_id: str | None = None,
        limit: int | None = None,
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream(
            "admin.connectors.sync_jobs",
            {
                "provider": provider,
                "collection": collection,
                "source_id": source_id,
                "limit": limit,
            },
        )

    def backups(
        self, *, limit: int | None = None, offset: int | None = None
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream("admin.backups", {"limit": limit, "offset": offset})

    def access_overview(
        self, *, window_days: int | None = None
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream("admin.access.overview", {"window_days": window_days})

    def access_logs(
        self,
        *,
        action: str | None = None,
        actor_id: str | None = None,
        collection: str | None = None,
        method: str | None = None,
        path: str | None = None,
        status_family: str | None = None,
        success: bool | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream(
            "admin.access.logs",
            {
                "action": action,
                "actor_id": actor_id,
                "collection": collection,
                "method": method,
                "path": path,
                "status_family": status_family,
                "success": success,
                "limit": limit,
                "offset": offset,
            },
        )

    def audit(
        self,
        *,
        action: str | None = None,
        actor_id: str | None = None,
        resource_type: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream(
            "admin.audit",
            {
                "action": action,
                "actor_id": actor_id,
                "resource_type": resource_type,
                "limit": limit,
                "offset": offset,
            },
        )

    def usage(self, *, window_days: int | None = None) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream("admin.usage", {"window_days": window_days})

    def platform_stats(self) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream("admin.platform.stats")

    def platform_readiness(self) -> AsyncGenerator[AdminRealtimeEvent, None]:
        return self._stream("admin.platform.readiness")

    def custom(
        self, topic: str, params: dict[str, Any] | None = None
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        if not topic.startswith("admin."):
            raise ValueError("admin.realtime.custom topic must start with admin.")
        return self._stream(topic, params or {})

    async def _stream(
        self, topic: str, params: dict[str, Any] | None = None
    ) -> AsyncGenerator[AdminRealtimeEvent, None]:
        connection = RealtimeConnection(self._client)
        try:
            async for message in connection.subscribe(topic, params or {}):
                yield message
                if message["type"] == "complete":
                    return
        finally:
            await connection.close()

File: sdks/typescript/src/resources/admin/realtime.ts
import type { RequestClient } from "../../core.js";
import { BigRAGRealtimeConnection } from "../../realtime.js";
import type {
  AccessLogListResponse,
  AccessLogOverviewResponse,
  AdminRealtimeEvent,
  AuditLogListResponse,
  BatchStatusResponse,
  CollectionStatsResponse,
  Document,
  DocumentListResponse,
  PlatformStatsResponse,
  ReadinessResponse,
  S3SourceListResponse,
  S3SyncJobListResponse,
  UploadSession,
  UsageResponse,
} from "../../types/index.js";

export class AdminRealtimeResource {
  constructor(private readonly _client: RequestClient) {}

  documents(
    collection: string,
    options: {
      limit?: number;
      offset?: number;
      order?: "asc" | "desc";
      q?: string;
      sort?: "created_at" | "updated_at" | "filename" | "file_size" | "chunk_count" | "status";
      status?: string;
    } = {},
  ): AsyncGenerator<AdminRealtimeEvent<DocumentListResponse>> {
    return this._stream("admin.collections.documents", {
      collection,
      limit: options.limit,
      offset: options.offset,
      order: options.order,
      q: options.q,
      sort: options.sort,
      status: options.status,
    });
  }

  documentBatchStatus(
    collection: string,
    documentIds: string[],
  ): AsyncGenerator<AdminRealtimeEvent<BatchStatusResponse>> {
    return this._stream("admin.collections.documents.batch_status", {
      collection,
      document_ids: documentIds,
    });
  }

  document(collection: string, documentId: string): AsyncGenerator<AdminRealtimeEvent<Document>> {
    return this._stream("admin.collections.documents.detail", {
      collection,
      document_id: documentId,
    });
  }

  uploadSession(
    collection: string,
    sessionId: string,
  ): AsyncGenerator<AdminRealtimeEvent<UploadSession>> {
    return this._stream("admin.collections.upload_session", { collection, session_id: sessionId });
  }

  collectionStats(collection: string): AsyncGenerator<AdminRealtimeEvent<CollectionStatsResponse>> {
    return this._stream("admin.collections.stats", { collection });
  }

  connectorSources(
    provider: string,
    options: { collection?: string } = {},
  ): AsyncGenerator<AdminRealtimeEvent<S3SourceListResponse>> {
    return this._stream("admin.connectors.sources", {
      provider,
      collection: options.collection,
    });
  }

  connectorSyncJobs(
    provider: string,
    options: { collection?: string; sourceId?: string; limit?: number } = {},
  ): AsyncGenerator<AdminRealtimeEvent<S3SyncJobListResponse>> {
    return this._stream("admin.connectors.sync_jobs", {
      provider,
      collection: options.collection,
      source_id: options.sourceId,
      limit: options.limit,
    });
  }

  accessOverview(
    options: { windowDays?: number } = {},
  ): AsyncGenerator<AdminRealtimeEvent<AccessLogOverviewResponse>> {
    return this._stream("admin.access.overview", {
      window_days: options.windowDays,
    });
  }

  accessLogs(
    options: {
      action?: string;
      actorId?: string;
      collection?: string;
      method?: string;
      path?: string;
      statusFamily?: string;
      success?: boolean;
      limit?: number;
      offset?: number;
    } = {},
  ): AsyncGenerator<AdminRealtimeEvent<AccessLogListResponse>> {
    return this._stream("admin.access.logs", {
      action: options.action,
      actor_id: options.actorId,
      collection: options.collection,
      method: options.method,
      path: options.path,
      status_family: options.statusFamily,
      success: options.success,
      limit: options.limit,
      offset: options.offset,
    });
  }

  audit(
    options: {
      action?: string;
      actorId?: string;
      resourceType?: string;
      limit?: number;
      offset?: number;
    } = {},
  ): AsyncGenerator<AdminRealtimeEvent<AuditLogListResponse>> {
    return this._stream("admin.audit", {
      action: options.action,
      actor_id: options.actorId,
      resource_type: options.resourceType,
      limit: options.limit,
      offset: options.offset,
    });
  }

  usage(options: { windowDays?: number } = {}): AsyncGenerator<AdminRealtimeEvent<UsageResponse>> {
    return this._stream("admin.usage", {
      window_days: options.windowDays,
    });
  }

  platformStats(): AsyncGenerator<AdminRealtimeEvent<PlatformStatsResponse>> {
    return this._stream("admin.platform.stats");
  }

  platformReadiness(): AsyncGenerator<AdminRealtimeEvent<ReadinessResponse>> {
    return this._stream("admin.platform.readiness");
  }

  custom<T = unknown>(
    topic: string,
    params: Record<string, string | number | boolean | string[] | undefined> = {},
  ): AsyncGenerator<AdminRealtimeEvent<T>> {
    if (!topic.startsWith("admin.")) {
      throw new Error("admin.realtime.custom topic must start with admin.");
    }
    return this._stream(topic, params);
  }

  private async *_stream<T>(
    topic: string,
    params: Record<string, string | number | boolean | string[] | undefined> = {},
  ): AsyncGenerator<AdminRealtimeEvent<T>> {
    const connection = new BigRAGRealtimeConnection(this._client);
    try {
      for await (const message of connection.subscribe<T>(topic, params)) {
        yield message as AdminRealtimeEvent<T>;
        if (message.type === "complete") return;
      }
    } finally {
      await connection.close();
    }
  }
}
